### PR TITLE
Fix padding calculation, closes #45

### LINKED ISSUE #45: Invalid blobs generated
I'm working on an application that sends blobs for controlling some other software, and ran into an issue in which this library generates invalid blobs. Sample code to recreate issue:

```
package main

import (
	"fmt"
	"github.com/hypebeast/go-osc/osc"
)

func main() {
	ip := "localhost"
	port := 7770
	client := osc.NewClient(ip, port)
	message := osc.NewMessage("/dmx/universe/0")
	b := make([]byte, 512)
	for i := 0; i < 512; i++ {
		b[i] = byte(i / 2)
	}
	message.Append(b)
	err := client.Send(message)
	fmt.Println(err)
	fmt.Println(message.Address)
	fmt.Println(message.Arguments)
}
```

OSCDump reports the following error:
liblo server error 9911 in path /dmx/universe/0: Invalid message received

If, however, I change the size of slice b to 511, everything works as intended:

```
package main

import (
	"fmt"
	"github.com/hypebeast/go-osc/osc"
)

func main() {
	ip := "localhost"
	port := 7770
	client := osc.NewClient(ip, port)
	message := osc.NewMessage("/dmx/universe/0")
	b := make([]byte, 511)
	for i := 0; i < 511; i++ {
		b[i] = byte(i / 2)
	}
	message.Append(b)
	err := client.Send(message)
	fmt.Println(err)
	fmt.Println(message.Address)
	fmt.Println(message.Arguments)
}
```
e3691c7e.717f8444 /dmx/universe/0 b [511 byte blob]

If there's any other useful information I can provide, let me know.

### PR DESCRIPTION
**Pull request by peternewman**
Fixes #45 

~Untested, but maths checked externally~

~Could probably do with some tests to catch such issues in future.~

### CODE REVIEW

File: osc/osc.go [hunk 1]
```diff
@@ -959,7 +959,7 @@ func readPaddedString(reader *bufio.Reader) (string, int, error) {
 	str = str[:len(str)-1]
```
Comment (chabad360) [inline osc/osc.go]: This line is the source of the issues.

Basically, OSC uses null terminated strings, which means that the delimiter is part of the calculation here.
> **OSC-string**
>  A sequence of non-null ASCII characters followed by a null, followed by 0-3 additional null characters to make the total number of bits a multiple of 32.

This means that a 4-byte string would actually be 5 bytes, in which case it's padded by an additional 3 bytes. Whereas a 3-byte string would only by 4 bytes, hence not padded at all.

Basically: Line #971 should become `return str[len(str)-1], n, nil`.

See my comments on the tests that you added.

File: osc/osc_test.go [hunk 2]
```diff
@@ -316,7 +316,9 @@ func TestReadPaddedString(t *testing.T) {
 		s   string // resulting string
 	}{
 		{[]byte{'t', 'e', 's', 't', 's', 't', 'r', 'i', 'n', 'g', 0, 0}, 12, "teststring"},
-		{[]byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8, "test"},
+		{[]byte{'t', 'e', 's', 't', 'e', 'r', 's', 0}, 8, "testers"},
+		{[]byte{'t', 'e', 's', 't', 's', 0, 0, 0}, 8, "tests"},
+		{[]byte{'t', 'e', 's', 't'}, 4, "test"},
```
Comment (chabad360) [inline osc/osc_test.go:321]: Replace these two with:

```go
{[]byte{'t', 'e', 's', 0, 0, 0, 0, 0}, 4, "tes", nil}, // OSC uses null terminated strings
{[]byte{'t', 'e', 's', 't'}, 0, "", io.EOF}, // if there is no null byte at the end, it doesn't work.
```

File: osc/osc.go [hunk 3]
```diff
@@ -956,10 +956,10 @@ func readPaddedString(reader *bufio.Reader) (string, int, error) {
 
 	// Remove the string delimiter, in order to calculate the right amount
 	// of padding bytes
-	str = str[:len(str)-1]
+	// str = str[:len(str)-1]
```
Comment (chabad360) [inline osc/osc.go:959]: Make sure to return `str[:len(str)-1]`

**Review by chabad360 — CHANGES_REQUESTED** [review]
See the comments regarding the string delimiter

File: osc/osc.go (hunk 1 above)
Comment (peternewman) [reply osc/osc.go]: Thanks @chabad360 ! I was getting close but unfortunately the OSC spec examples page seems to be gone forever! :cry:

File: osc/osc_test.go (hunk 2 above)
Comment (peternewman) [reply osc/osc_test.go]: > ```go
> {[]byte{'t', 'e', 's', 0, 0, 0, 0, 0}, 4, "tes", nil}, // OSC uses null terminated strings
> ```

Surely this one should be:
> ```go
> {[]byte{'t', 'e', 's', 0}, 4, "tes", nil}, // OSC uses null terminated strings
> ```

Which is what I've committed, as the null terminator takes the total string length to 4 bytes so no padding required.

Comment (chabad360) [inline osc/osc_test.go]: This test currently fails.

**Review by chabad360 — APPROVED** [review]
LGTM.

**Review by chabad360 — CHANGES_REQUESTED** [review]
Oops, forgot one more thing: `writePaddedString()` needs to be modified as well, otherwise we still have this issue.

**Comment by peternewman** [comment]
> You should probably update the PR text to at least say `fixes #45`, this way it'll be automatically closed. Also, make sure to mark it ready for review.

I thought it worked in the title too, but added to both to be sure.

**Comment by peternewman** [comment]
> Oops, forgot one more thing: [`writePaddedString()`](https://github.com/peternewman/go-osc/blob/patch-1/osc/osc.go#L973-L993) needs to be modified as well, otherwise we still have this issue.

I don't see where @chabad360 . It uses the count from writeString which presumably already includes the null terminator. Or do you mean we need to ensure one is added?

File: osc/osc_test.go (hunk 2 above)
Comment (peternewman) [reply osc/osc_test.go]: Travis is now green.

Comment (peternewman) [reply osc/osc_test.go]: Travis seems happy with my correction too @chabad360 unless you think I've missed something?

**Comment by chabad360** [comment]
We need to ensure one is added.

**Comment by peternewman** [comment]
> We need to ensure one is added.
> 
> I would just add
> 
> ```go
> buf.WriteByte(0)
> n += 1
> ```
> 
> right after `buf.WriteString()` (worked for me).

Won't that get a double one if the string already has one (or is that impossible/guaranteed in Go)? Consider:
`'t', 'e', 's', '\0'`
You'd end up sending:
`'t', 'e', 's', 0, 0, 0, 0, 0`

**Comment by chabad360** [comment]
Unless you specifically craft a string to include a null byte `string([]byte{0})`, a string shouldn't contain one. However, because it's not simple to add a null byte efficiently (nor is it idiomatic), it's safer to assume it's not there. Additionally, all the tests I ran (and some analysis) showed that the library doesn't insert any null bytes.

File: osc/osc_test.go (hunk 2 above)
Comment (chabad360) [reply osc/osc_test.go]: Reason why I had included the extra 4 bytes, was just to ensure that they are ignored.

File: osc/osc_test.go [hunk 4]
```diff
@@ -314,14 +315,20 @@ func TestReadPaddedString(t *testing.T) {
...
 		s   string // resulting string
+		e   error  // expected error
 	}{
-		{[]byte{'t', 'e', 's', 't', 's', 't', 'r', 'i', 'n', 'g', 0, 0}, 12, "teststring"},
-		{[]byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8, "test"},
+		{[]byte{'t', 'e', 's', 't', 'S', 't', 'r', 'i', 'n', 'g', 0, 0}, 12, "testString", nil},
+		{[]byte{'t', 'e', 's', 't', 'e', 'r', 's', 0}, 8, "testers", nil},
+		{[]byte{'t', 'e', 's', 't', 's', 0, 0, 0}, 8, "tests", nil},
+		{[]byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8, "test", nil},
+		{[]byte{}, 0, "", io.EOF},
```
Comment (peternewman) [inline osc/osc_test.go:324]: This currently returns EOF, I think something suggested there may be no string, so should TestReadPaddedString catch this?

Comment (chabad360) [reply osc/osc_test.go:324]: I don't see why it shouldn't, that's the expected behavior, so technically, a deviation from that is bad.

Comment (peternewman) [reply osc/osc_test.go:324]: So we SHOULD expect this to fail with an EOL?

File: osc/osc.go [hunk 5]
```diff
@@ -977,8 +977,14 @@ func writePaddedString(str string, buf *bytes.Buffer) (int, error) {
 		return 0, err
 	}
 
+	// Add a null terminator if not already present
+	if str[:len(str)-1] != 0 {
```
Comment (chabad360) [inline osc/osc.go:981]: That should be:

```go
if str[len(str)-1] != 0 {
```

File: osc/osc_test.go (hunk 4 above)
Comment (chabad360) [reply osc/osc_test.go:324]: Yes, this is expected behavior.

File: osc/osc.go (hunk 5 above)
Comment (peternewman) [reply osc/osc.go]: Which is exactly what I wrote in the playground, and then failed to write here :facepalm:

File: osc/osc_test.go [hunk 6]
```diff
@@ -333,46 +340,75 @@ func TestReadPaddedString(t *testing.T) {
...
+		s   string // string
+		buf []byte // resulting buffer
+		n   int    // bytes expected
+	}{
+		{"testString", []byte{'t', 'e', 's', 't', 'S', 't', 'r', 'i', 'n', 'g', 0, 0}, 12},
+		{"testers", []byte{'t', 'e', 's', 't', 'e', 'r', 's', 0}, 8},
+		{"tests", []byte{'t', 'e', 's', 't', 's', 0, 0, 0}, 8},
+		{"test", []byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8},
+		{"tes", []byte{'t', 'e', 's', 0}, 4},
+		{"tes\0", []byte{'t', 'e', 's', 0}, 4}, // Don't add a second null terminator if one is already present
```
Comment (chabad360) [inline osc/osc_test.go:353]: these two appear to be the same thing, perhaps the first one shouldn't have the null byte?

File: osc/osc_test.go (hunk 2 above)
Comment (peternewman) [reply osc/osc_test.go]: Makes sense, but the length should still be four as it's output not input length. I've added it and another couple of separate tests.

File: osc/osc_test.go (hunk 6 above)
Comment (peternewman) [reply osc/osc_test.go]: Input is on the left, so only line 353 feed in a pre-existing null (assuming Go strings don't normally have them), but both should result in identical output.

File: osc/osc_test.go [hunk 7]
```diff
@@ -314,14 +315,22 @@ func TestReadPaddedString(t *testing.T) {
...
 	}{
-		{[]byte{'t', 'e', 's', 't', 's', 't', 'r', 'i', 'n', 'g', 0, 0}, 12, "teststring"},
-		{[]byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8, "test"},
+		{[]byte{'t', 'e', 's', 't', 'S', 't', 'r', 'i', 'n', 'g', 0, 0}, 12, "testString", nil},
+		{[]byte{'t', 'e', 's', 't', 'e', 'r', 's', 0}, 8, "testers", nil},
+		{[]byte{'t', 'e', 's', 't', 's', 0, 0, 0}, 8, "tests", nil},
+		{[]byte{'t', 'e', 's', 't', 0, 0, 0, 0}, 8, "test", nil},
+		{[]byte{}, 0, "", io.EOF},
+		{[]byte{'t', 'e', 's', 0}, 4, "tes", nil},             // OSC uses null terminated strings
+		{[]byte{'t', 'e', 's', 0, 0, 0, 0, 0}, 4, "tes", nil}, // Additional nulls should be ignored
```
Comment (chabad360) [inline osc/osc_test.go:326]: This test will fail.
I would argue that this test shouldn't be included, unless you include some extra checking to ensure that it only writes that proper amount (which I don't think is necessary).

I think I'm going to look into writing fuzz tests for this.

File: osc/osc_test.go (hunk 6 above)
Comment (chabad360) [reply osc/osc_test.go]: ah, yes, that makes sense.

**Review by chabad360 — APPROVED** [review]
Alright, LGTM. @hypebeast?

File: osc/osc.go [hunk 8]
```diff
@@ -968,20 +964,30 @@ func readPaddedString(reader *bufio.Reader) (string, int, error) {
...
+	// Strip off the string delimiter
+	return str[:len(str)-1], n, nil
 }
 
 // writePaddedString writes a string with padding bytes to the a buffer.
 // Returns, the number of written bytes and an error if any.
 func writePaddedString(str string, buf *bytes.Buffer) (int, error) {
+	// Truncate at the first null, just in case there is more than one present
+	nullIndex := strings.Index(str, "\x00")
+	if (nullIndex > 0) {
```
Comment (hypebeast) [inline osc/osc.go:976]: Gofmt ist complaining about the parenthesis. Therefore the build fails.

**Review by hypebeast — COMMENTED** [review]
Very nice. This PR looks good to me.

Thanks a lot for it. As soon as the build succeeds I'll merge it.

File: osc/osc.go (hunk 8 above)
Comment (peternewman) [reply osc/osc.go]: Done

**Comment by peternewman** [comment]
> Very nice. This PR looks good to me.

Thanks @hypebeast . It seems to works for all the test cases @chabad360 came up with and a few edge cases I conjured up too.

> Thanks a lot for it.

No worries, not quite the easy fix I first thought, but hopefully should make things better and more standardised all round.

**Comment by peternewman** [comment]
This now passes aside from https://travis-ci.org/github/hypebeast/go-osc/jobs/764014122#L275-L279 which is also present in master @hypebeast .

Looks like this could fix it, but a bit outside my experience:
https://github.com/ipfs/go-ipfs/issues/2043

**Comment by chabad360** [comment]
I would say that issue is probably outside the scope of this PR.

**Comment by hypebeast** [comment]
Yes, this is outside of this PR. Let me see if I can fix it ;)

Thanks again for your work @peternewman and @chabad360
